### PR TITLE
Filesystem: refactor, improvements, add tests

### PR DESCRIPTION
**Pull request by pilou-**
##### SUMMARY
- add integration tests
    - creation, resizing (not for XFS) and overwriting are tested
    - all filesystem but LVM are tested
- improvements:
  - module documentation (requirements, add a warning)
  - explicitly fails when resizing isn't supported
  - handle btrfs < 0.20-rc1
  - device: allow to use regular file
- Fix PEP8.

##### ISSUE TYPE
 - Bugfix Pull Request
 - Docs Pull Request

##### COMPONENT NAME
filesystem

##### ANSIBLE VERSION

```
ansible-playbook 2.5.0 (devel e3f1198a67) last updated 2017/11/01 22:49:06 (GMT +200)
```

##### ADDITIONAL INFORMATION
Related: #24356

### CODE REVIEW

File: lib/ansible/modules/system/filesystem.py
Comment (Akasurde) [inline lib/ansible/modules/system/filesystem.py]: I am wondering why 'btrfs' block is removed.

Comment (pilou-) [reply lib/ansible/modules/system/filesystem.py]: Because the current code doesn't support filesystem resizing when the filesystem is `btrfs`. The size in bytes of a `btrfs` filesystem is not 1.

File: lib/ansible/modules/system/filesystem.py
Comment (mscherer) [inline lib/ansible/modules/system/filesystem.py]: Shouldn't it fail rather than just exist without error ?

File: lib/ansible/modules/system/filesystem.py
Comment (mscherer) [inline lib/ansible/modules/system/filesystem.py]: It might be better to say "all other filesystems do not support resizing". (so adding a filesystem mean that we have one less place to modify)

**Comment by mscherer** [comment]
If we reach agreement for the exit_json vs fail_json stuff when resizing, that's good for me.

**Comment by pilou-** [comment]
> Shouldn't it fail rather than just exist without error ?

you're right, it must fail, thanks.

File: lib/ansible/modules/system/filesystem.py
Comment (pilou-) [reply lib/ansible/modules/system/filesystem.py]: Done (list of unsupported FS replaced by list of supported FS).

**Comment by mscherer** [comment]
shipit

**Review by pchauncey — APPROVED** [review]
shipit

**Comment by pilou-** [comment]
bot_status

**Comment by pilou-** [comment]
The bot labeled this pull-request with `stale_ci` because the CI check is older than one week. Since a few days, this label [prevents potential automerge](https://github.com/ansible/ansibullbot/pull/661/commits/58364d7e229528cef069293e636fbf7e192c88d2). A rebase is required in order to trigger the CI, then this pull-request will become compatible with the automerge workflow. `shipit` approval added before this rebase will be ignored by the bot.

**Comment by pilou-** [comment]
Pull request updated (PEP8 errors fixed) and rebased.

@misc: please, could you comment with `shipit` again ?

@Akasurde: please, could you review it again ?

**Comment by mscherer** [comment]
shipit

**Comment by pilou-** [comment]
@azaghal @tdtrask (as randomly selected maintainers of `system` namespace): could you please review this pull request ? At least two `shipit` comments (from maintainers or namespace maintainers or core team members) are required in order to [automerge](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#community) this pull request.

File: lib/ansible/modules/system/filesystem.py
Comment (azaghal) [inline lib/ansible/modules/system/filesystem.py]: More of a nitpicking than anything else - but wouldn't it be better to perform check for resize operation + btrfs within the `main()` function? Would be better context, primarily because this function call is specifically about getting the file system size.

File: lib/ansible/modules/system/filesystem.py
Comment (azaghal) [reply lib/ansible/modules/system/filesystem.py]: Would it be hard to implement proper support for it? I have not used btrfs myself, so no idea if this would be very complicated or not?

**Comment by pilou-** [comment]
Rebased (conflict resolved).

**Comment by mscherer** [comment]
shipit

File: lib/ansible/modules/system/filesystem.py
Comment (azaghal) [inline lib/ansible/modules/system/filesystem.py]: This can be considered a bit of nitpicking perhaps, and other reviewers may think it does not matter much, but - wouldn't it be better to move this particular failure condition + message outside of this function and somewhere into the `main()` as part of verifying the parameters?

File: lib/ansible/modules/system/filesystem.py
Comment (azaghal) [inline lib/ansible/modules/system/filesystem.py]: A bit related to my upper comment about moving the failure message for module not supporting resize of particular filesystem, it might make some sense to actually set the `grow` and/or `grow_flag` to `None`, and then essentially check for this value when user requests filesystem resizing to see if filesystem resizing is supported.

While at it, is it just me, or the 'grow_flag' is actually not used at all anywhere in the code? This is not related to your changes per-se, just something I tried to figure out quickly.

**Review by azaghal — CHANGES_REQUESTED** [review]
Primarily nitpicking around some small code organisation improvements.

File: lib/ansible/modules/system/filesystem.py
Comment (pilou-) [reply lib/ansible/modules/system/filesystem.py]: > Would it be hard to implement proper support for it? I have not used btrfs myself, so no idea if this would be very complicated or not?

It appears that `btrfs filesystem resize /mounted/path`. I propose that I implement that in another PR.

> wouldn't it be better to move this particular failure condition + message outside of this function and somewhere into the main() as part of verifying the parameters?

This is now tested into the `main` method.

File: lib/ansible/modules/system/filesystem.py
Comment (pilou-) [reply lib/ansible/modules/system/filesystem.py]: This is now tested into the main method.

**Comment by pilou-** [comment]
@mscherer @azaghal  @pchauncey @Akasurde please, could you review this updated pull-request ?

Some refactor has been done and integration tests have been added.

File: test/integration/targets/filesystem/tasks/overwrite_another_fs.yml
Comment (mscherer) [inline test/integration/targets/filesystem/tasks/overwrite_another_fs.yml:6]: Wouldn't it be more future-proof to use ansible_system == 'Linux' ? How well would it work on others BSD and/or windows ?

File: test/integration/targets/filesystem/tasks/setup.yml
Comment (mscherer) [inline test/integration/targets/filesystem/tasks/setup.yml]: I am not sure to understand that conditional. Why do you add dosfstools if that's excluded each time ?

**Comment by mscherer** [comment]
shipit

File: test/integration/targets/filesystem/tasks/setup.yml
Comment (pilou-) [reply test/integration/targets/filesystem/tasks/setup.yml]: That's not excluded when `ansible_system != 'FreeBSD'` is `True`.

File: test/integration/targets/filesystem/tasks/overwrite_another_fs.yml
Comment (pilou-) [reply test/integration/targets/filesystem/tasks/overwrite_another_fs.yml:6]: Due to `aliases` file it won't be executed on Windows but it could be problematic with others BSD. Updated.

**Comment by pilou-** [comment]
Updated:
- fix conflicts
- handle @mscherer comment

File: lib/ansible/modules/system/filesystem.py
Comment (bcoca) [inline lib/ansible/modules/system/filesystem.py]: i would rephrase: `Makes a file system` or `Makes file systems`

File: lib/ansible/modules/system/filesystem.py
Comment (bcoca) [inline lib/ansible/modules/system/filesystem.py]: `This module creates a file system`

File: lib/ansible/modules/system/filesystem.py
Comment (bcoca) [reply lib/ansible/modules/system/filesystem.py]: I'm not sure which one is more valid `filesystem` or `file system` but in any case it should be either plural or have an 'a'

File: lib/ansible/modules/system/filesystem.py
Comment (bcoca) [inline lib/ansible/modules/system/filesystem.py]: As much as i doubt which form is better above .. camel case is not in the running at all

File: lib/ansible/modules/system/filesystem.py
Comment (bcoca) [inline lib/ansible/modules/system/filesystem.py]: since its  not really a 'regular file' I would change it to 'Target path to device or image file' or similar

File: lib/ansible/modules/system/filesystem.py
Comment (bcoca) [inline lib/ansible/modules/system/filesystem.py]: the problem already existed, but this line really belongs in 'requirements' instead of notes section.

**Review by bcoca — CHANGES_REQUESTED** [review]
minor doc nitpicking

File: lib/ansible/modules/system/filesystem.py
Comment (pilou-) [reply lib/ansible/modules/system/filesystem.py]: > I'm not sure which one is more valid filesystem or file system

Used terms were not consistent (`filessytem`, `filesytem`, `file system`), I followed [man pages conventions](http://man7.org/linux/man-pages/man7/man-pages.7.html#STYLE_GUIDE) and choose `filesystem`.

> in any case it should be either plural or have an 'a'

Done

File: lib/ansible/modules/system/filesystem.py
Comment (pilou-) [reply lib/ansible/modules/system/filesystem.py]: Done (using: `This module creates a filesystem.`)

File: lib/ansible/modules/system/filesystem.py
Comment (pilou-) [reply lib/ansible/modules/system/filesystem.py]: Done

File: lib/ansible/modules/system/filesystem.py
Comment (pilou-) [reply lib/ansible/modules/system/filesystem.py]: Done

File: lib/ansible/modules/system/filesystem.py
Comment (pilou-) [reply lib/ansible/modules/system/filesystem.py]: Done

**Comment by pilou-** [comment]
@bcoca pull-request updated, thanks !

**Comment by gundalow** [comment]
@mscherer @azaghal @pchauncey @Akasurde @bcoca I believe all the review comments have been addressed, could we please get a final review from you on this. Would be great to get this into 2.5.

Thanks in advance

**Review by pchauncey — APPROVED** [review]
shipit

**Comment by pilou-** [comment]
ready_for_review (testing bot behaviour :)

**Comment by mscherer** [comment]
shipit

**Comment by abadger** [comment]
rebuild_merge

**Comment by ghost** [comment]
Too list and define a class for specific file system type can be useful. But for now this excludes swap.

I think this is a kind of regression.

**Pull request merged**
